Integrate @let into the runtime

crisbeto opened the pull request:
Includes the following changes that integrate the new `@let` syntax into the runtime. 

### refactor(core): integrate let instructions into the runtime
Adds the implementation of the following new instructions:
* `declareLet` - creation-time instruction that initializes the slot for a let declaration.
* `storeLet` - update-time instruction that stores the current value of a let declaration.
* `readContextLet` - instruction that reads the stored value of a let declaration from a different view.

On top of the instructions, it also introduces a new `LetDeclaration` TNode type. The new TNode is nececessary for DI to work correctly in pipes inside the let expression, as well as for proper hydration support.

### refactor(core): support let declarations during hydration
Updates the hydration logic to account for the fact that let declarations don't create a DOM node.

crisbeto committed:
refactor(core): integrate let instructions into the runtime

Adds the implementation of the following new instructions:
* `declareLet` - creation-time instruction that initializes the slot for a let declaration.
* `storeLet` - update-time instruction that stores the current value of a let declaration.
* `readContextLet` - instruction that reads the stored value of a let declaration from a different view.

On top of the instructions, it also introduces a new `LetDeclaration` TNode type.
The new TNode is nececessary for DI to work correctly in pipes inside the let expression,
as well as for proper hydration support.

crisbeto commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -742,6 +748,16 @@ export interface TProjectionNode extends TNode {
 }
 
+/** Static data for a `@let` declaration. */
+export interface TLetDeclarationNode extends TNode {
```
The context why we need this node is that pipes use the node injector for DI. The node injector only works if there's a `TNode` so we need one so that things like the `async` pipe work within `@let`.

AndrewKushnir replied:
This is a great comment that we can capture in the JSDoc in the code 👍

AndrewKushnir commented on packages/core/src/render3/instructions/let_declaration.ts:
```diff
@@ -6,36 +6,67 @@
+  if (value === UNINITIALIZED_LET) {
+    throw new RuntimeError(
+      RuntimeErrorCode.UNINITIALIZED_LET_ACCESS,
+      'Attempting to access a @let declaration whose value is not available yet',
```
We can add `ngDevMode` here to tree-shake-away the error message:
```suggestion
      ngDevMode && 'Attempting to access a @let declaration whose value is not available yet',
```

AndrewKushnir commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -525,24 +525,25 @@ function serializeLView(lView: LView, context: HydrationContext): SerializedView
+        // thus it has no DOM elements associated with it, so the **next sibling**
+        // node would not be able to find an anchor. In this case, use full path instead.
         let nextTNode = tNode.next;
         // Skip over all `<ng-content>` slots in a row.
```
```suggestion
        // Skip over all `<ng-content>` slots and `@let` declarations.
```

AndrewKushnir reviewed (commented):
@crisbeto looks great, just left a few minor comments.

crisbeto committed:
refactor(core): support let declarations during hydration

Updates the hydration logic to account for the fact that let declarations don't create a DOM node.

crisbeto commented on packages/core/src/render3/instructions/let_declaration.ts:
```diff
@@ -6,36 +6,67 @@
+  if (value === UNINITIALIZED_LET) {
+    throw new RuntimeError(
+      RuntimeErrorCode.UNINITIALIZED_LET_ACCESS,
+      'Attempting to access a @let declaration whose value is not available yet',
```
We currently don't pass the name to the instructions. It might make sense in the future to do it so dev tools can visualize the values, but I left it out for now since this error is fairly hard to hit.

crisbeto commented:
Thank you for the feedback @AndrewKushnir, it has been addressed.

AndrewKushnir reviewed (approved):
@crisbeto looks great, thanks! 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

dylhunn commented:
This PR was merged into the repository by commit 7dfe302f4797380562b19d6271295140a03396be.

The changes were merged into the following branches: main, 18.0.x
